x/stakeibc/keeper/validator_selection_test.go

```diff
@@ -0,0 +1,270 @@
+		s.Require().Equal(expected.ValidatorDstAddress, actual.ValidatorDstAddress, "message dst validator, index %d", i)
+	}
+}
+
+func (s *KeeperTestSuite) TestGetRebalanceICAMessages_EvenNumberValidators() {
+	// Build up deltas for each validator, i.e. how much each validator needs to change by
+	validatorDeltas := []keeper.RebalanceValidatorDelegationChange{
+		// Overweight validators - they should lose some of their stake
+		{ValidatorAddress: "val1", Delta: sdkmath.NewInt(-21)}, // 15 to val10, 6 to val9
+		{ValidatorAddress: "val2", Delta: sdkmath.NewInt(-19)}, // 6 to val9, 10 to val8, 3 to val7
+		{ValidatorAddress: "val3", Delta: sdkmath.NewInt(-13)}, // 3 to val7, 5 to val6, 4 to val5, 1 to val4
+
+		// Underweight validators - they should gain stake
+		{ValidatorAddress: "val4", Delta: sdkmath.NewInt(1)},   // 1 from val3
+		{ValidatorAddress: "val5", Delta: sdkmath.NewInt(4)},   // 4 from val3
+		{ValidatorAddress: "val6", Delta: sdkmath.NewInt(5)},   // 5 from val3
+		{ValidatorAddress: "val7", Delta: sdkmath.NewInt(6)},   // 3 from val2, 3 from val3
+		{ValidatorAddress: "val8", Delta: sdkmath.NewInt(10)},  // 10 from val2
+		{ValidatorAddress: "val9", Delta: sdkmath.NewInt(12)},  // 6 from val1, 6 from val2
+		{ValidatorAddress: "val10", Delta: sdkmath.NewInt(15)}, // 15 from val1
+	}
+
+	// Build up the expected messages, moving across the list above
+	expectedRebalancings := []types.Rebalancing{
+		{SrcValidator: "val1", DstValidator: "val10", Amt: sdkmath.NewInt(15)}, // 15 from val1 to val10
+		{SrcValidator: "val1", DstValidator: "val9", Amt: sdkmath.NewInt(6)},   //  6 from val1 to val9
+
+		{SrcValidator: "val2", DstValidator: "val9", Amt: sdkmath.NewInt(6)},  //  6 from val2 to val9
+		{SrcValidator: "val2", DstValidator: "val8", Amt: sdkmath.NewInt(10)}, // 10 from val2 to val8
+		{SrcValidator: "val2", DstValidator: "val7", Amt: sdkmath.NewInt(3)},  //  3 from val2 to val7
+
+		{SrcValidator: "val3", DstValidator: "val7", Amt: sdkmath.NewInt(3)}, // 3 from val3 to val7
+		{SrcValidator: "val3", DstValidator: "val6", Amt: sdkmath.NewInt(5)}, // 5 from val3 to val6
+		{SrcValidator: "val3", DstValidator: "val5", Amt: sdkmath.NewInt(4)}, // 4 from val3 to val5
+		{SrcValidator: "val3", DstValidator: "val4", Amt: sdkmath.NewInt(1)}, // 1 from val3 to val4
+	}
+
+	s.checkRebalanceICAMessages(validatorDeltas, expectedRebalancings)
+}
+
+func (s *KeeperTestSuite) TestGetRebalanceICAMessages_OddNumberValidators() {
+	// Build up deltas for each validator, i.e. how much each validator needs to change by
+	validatorDeltas := []keeper.RebalanceValidatorDelegationChange{
+		// Overweight validators - they should lose some of their stake
+		{ValidatorAddress: "val1", Delta: sdkmath.NewInt(-15)}, // 15 to val9
+		{ValidatorAddress: "val2", Delta: sdkmath.NewInt(-12)}, // 6 to val9, 6 to val8
+		{ValidatorAddress: "val3", Delta: sdkmath.NewInt(-9)},  // 9 to val8
```

Comment:
nit, can we make `val3` and `val4` -8 and -8? Just to add one more edge-case

Reply:
I want to this in there cause this is the edge case from our earlier comments where all the delta from one val gets moved to the other val and they both increment indicies.

Is the case you wanna add one where two validator's have the same stake? Cause I can add that case in lower in this test!

Reply:
added another case, lmk if it resolves it for you

Reply:
Ah gotcha that makes sense, yeah agreed that resolves it for me! Thanks for making that change!